cdc: refactor to replace map of changefeed targets with array

HonoreDB opened the pull request:
Necessary intermediate step towards changefeeds on things other than full tables.
Previously, changefeed targets were serialized as a map of TableID->StatementTimeName.
This keeps that for now for compatibility but adds in a list of target specifications
that's more extendable, and changes the API for everything that will eventually need
to care (for example, an encoder will need to know that it's encoding a [column family](https://github.com/cockroachdb/cockroach/issues/28667)).

Still needs a test with the old-style payload.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/75430)
<!-- Reviewable:end -->

miretskiy commented on pkg/jobs/jobspb/jobs.proto:
```diff
@@ -695,40 +695,53 @@ message SchemaChangeGCProgress {
 }
 
-message ChangefeedTarget {
+message ChangefeedTargetTable {
```
I'm curious: why do we need to change ChangefeedTarget and introduce 2 new types? Couldn't we extend
ChangefeedTarget message itself?  Say, add the TargetType (defulat == 0 -- so , same as now), add e.g. family_name if target type != 0?

HonoreDB replied:
My reasoning here is that we won't be able to put things like family_name into the ChangefeedTargets (note plural) proto, since it's an ID -> Target map and column families don't have globally unique ids, and the map needs to stay for now for inter-version compatibility. So if we extend ChangefeedTarget, we end up with both a map and an array of ChangefeedTarget where the copies in the map are missing some data, which seems overall more confusing. We don't technically have to change ChangefeedTarget to ChangefeedTargetTable, I just wanted the name to stay accurate.

HonoreDB commented on pkg/jobs/jobspb/jobs.proto:
```diff
@@ -695,40 +695,53 @@ message SchemaChangeGCProgress {
+
+    // The primary index of the table with table_id descriptor id.
+    // Each column family gets its own record schema and events.
+    EACH_FAMILY = 1;
```
each_family is a feed on the entire table, but tolerating the existence of multiple column families and emitting them separately. column_family is a feed on a single, specified column family of a table. i.e. the first is "TABLE foo" after we turn off the column family check by default, and the second is "TABLE foo FAMILY bar".

HonoreDB commented on pkg/jobs/jobspb/jobs.proto:
```diff
@@ -695,40 +695,53 @@ message SchemaChangeGCProgress {
+  repeated ChangefeedTargetSpecification target_specifications = 8 [(gogoproto.nullable) = false];
 
   reserved 1, 2, 5;
+  reserved "targets";
```
Reusing a field name allegedly can cause bugs in some libraries, per the docs. Maybe nothing we use here, I don't know.

miretskiy reviewed (changes requested):
> Reviewed 3 of 20 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75430)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @gh-casper and @HonoreDB)

___
*[pkg/jobs/jobspb/jobs.proto, line 698 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75430#-Mu8j3bSBoVUX8VZ6xE5-r1-698:-MuklbJB3PCNClB2ZQ8n:b-9755fa) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeb0c0805bf9539e59f98f9d50d814597bd0876/pkg/jobs/jobspb/jobs.proto#L698)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

My reasoning here is that we won't be able to put things like family_name into the ChangefeedTargets (note plural) proto, since it's an ID -> Target map and column families don't have globally unique ids, and the map needs to stay for now for inter-version compatibility. So if we extend ChangefeedTarget, we end up with both a map and an array of ChangefeedTarget where the copies in the map are missing some data, which seems overall more confusing. We don't technically have to change ChangefeedTarget to ChangefeedTargetTable, I just wanted the name to stay accurate.
</blockquote></details>

Thank you.
___
*[pkg/jobs/jobspb/jobs.proto, line 710 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75430#-Mu8j3bSBoVUX8VZ6xE5-r1-710:-MuXT-AIDLuQt5zN9BSO:b9nea9n) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeb0c0805bf9539e59f98f9d50d814597bd0876/pkg/jobs/jobspb/jobs.proto#L710)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

each_family is a feed on the entire table, but tolerating the existence of multiple column families and emitting them separately. column_family is a feed on a single, specified column family of a table. i.e. the first is "TABLE foo" after we turn off the column family check by default, and the second is "TABLE foo FAMILY bar".
</blockquote></details>

Thanks for the explanation.
This might be a nit -- so, feel free to ignore... 
Some ideas for different names: Perhaps FULL_TABLE should be renamed to "PRIMARY_ONLY" (or PRIMARY_FAMILY -- to be consistent with other fields, or even better, PRIMARY_FAMILY_ONLY); 
EACH_FAMILY -> ALL_FAMILIEIS
COLUMN_FAMILY -> SPECIFIC_FAMILY?
___
*[pkg/jobs/jobspb/jobs.proto, line 744 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75430#-Mu8j3bSBoVUX8VZ6xE5-r1-744:-MukkpotCr8H7g04ixjT:bgrcgfe) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeb0c0805bf9539e59f98f9d50d814597bd0876/pkg/jobs/jobspb/jobs.proto#L744)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

Reusing a field name allegedly can cause bugs in some libraries, per the docs. Maybe nothing we use here, I don't know.
</blockquote></details>

I think it maybe more todo w/ json encoding?
___
*[pkg/jobs/jobspb/jobs.proto, line 733 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75430#-MuklZNQ75MFe658R4eq:-MuklZNQ75MFe658R4er:b-u0q9bq) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b2d95ffe4610e2ab253aa2557f0306ab8b11d63/pkg/jobs/jobspb/jobs.proto#L733)):*
> ```protobuf
>   // renames can be tolerated and derived topic names remain immutable.
>   //
>   map<uint32, ChangefeedTargetTable> tables = 6 [
> ```

Based on your explanation above (on why we are adding new types), it does appear that the plan is to get rid of this field at some point and replace those with target_specifications?

If so, what do you think about renaming ChangeFeedTargetTable -> DeprecatedChangeFeedTargetTable, and rename tables->deprecated_tables?

Do we need version gates for this?  Or is the intention to have this backportable for now and add version gate later?


<!-- Sent from Reviewable.io -->

HonoreDB reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75430)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @gh-casper, @HonoreDB, and @miretskiy)

___
*[pkg/jobs/jobspb/jobs.proto, line 710 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75430#-Mu8j3bSBoVUX8VZ6xE5-r1-710:-Muq8pWK84O4zDweraes:b-dxcbts) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeb0c0805bf9539e59f98f9d50d814597bd0876/pkg/jobs/jobspb/jobs.proto#L710)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Thanks for the explanation.
This might be a nit -- so, feel free to ignore... 
Some ideas for different names: Perhaps FULL_TABLE should be renamed to "PRIMARY_ONLY" (or PRIMARY_FAMILY -- to be consistent with other fields, or even better, PRIMARY_FAMILY_ONLY); 
EACH_FAMILY -> ALL_FAMILIEIS
COLUMN_FAMILY -> SPECIFIC_FAMILY?


</blockquote></details>

Makes sense.
___
*[pkg/jobs/jobspb/jobs.proto, line 744 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75430#-Mu8j3bSBoVUX8VZ6xE5-r1-744:-Muq8cGj1_qxwNr54xKn:bc9dbsz) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeb0c0805bf9539e59f98f9d50d814597bd0876/pkg/jobs/jobspb/jobs.proto#L744)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I think it maybe more todo w/ json encoding?
</blockquote></details>

Do you think it's something we don't need to worry about? I'd rather Target than TargetSpecification I think if it's not going to cause bugs.
___
*[pkg/jobs/jobspb/jobs.proto, line 733 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/75430#-MuklZNQ75MFe658R4eq:-Muq89lL0y_uWvMBne0n:bmz6niw) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b2d95ffe4610e2ab253aa2557f0306ab8b11d63/pkg/jobs/jobspb/jobs.proto#L733)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Based on your explanation above (on why we are adding new types), it does appear that the plan is to get rid of this field at some point and replace those with target_specifications?

If so, what do you think about renaming ChangeFeedTargetTable -> DeprecatedChangeFeedTargetTable, and rename tables->deprecated_tables?

Do we need version gates for this?  Or is the intention to have this backportable for now and add version gate later?
</blockquote></details>

Intention is to have this backportable for now, yeah. There is a small legit use for the table map since in some methods we only have a descriptor ID so this lets us get the statement time name more efficiently, so I'm not in a hurry to deprecate this but probably we will eventually.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75430)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @gh-casper, @HonoreDB, and @miretskiy)

___
*[pkg/jobs/jobspb/jobs.proto, line 744 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/75430#-Mu8j3bSBoVUX8VZ6xE5-r1-744:-MuqIPVsB6JdnzdAVXnu:b1ppqwx) ([raw file](https://github.com/cockroachdb/cockroach/blob/0aeb0c0805bf9539e59f98f9d50d814597bd0876/pkg/jobs/jobspb/jobs.proto#L744)):*
<details><summary><i>Previously, HonoreDB (Aaron Zinger) wrote…</i></summary><blockquote>

Do you think it's something we don't need to worry about? I'd rather Target than TargetSpecification I think if it's not going to cause bugs.
</blockquote></details>

It's okay.. just was trying to understand.


<!-- Sent from Reviewable.io -->

HonoreDB committed:
cdc: refactor to replace map of changefeed targets with array

Necessary intermediate step towards changefeeds on things other than full tables.
Previously, changefeed targets were serialized as a map of TableID->StatementTimeName.
This keeps that for now for compatibility but adds in a list of target specifications
that's more extendable, and changes the API for everything that will eventually need
to care (for example, an encoder will need to know that it's encoding a column family).

Tested manually that an old-style payload still works, and some existing unit tests
create an old-style ChangefeedDetails.

Release note: None

miretskiy commented:
@HonoreDB do you need another pass over this?

miretskiy reviewed (approved):
> Reviewed 1 of 20 files at r1, 2 of 4 files at r2, 9 of 15 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75430)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @gh-casper, @HonoreDB, and @miretskiy)

___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 968 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75430#-MwWoFrh47N63DdmBnhQ:-MwWoFrh47N63DdmBnhR:b-t9veiy) ([raw file](https://github.com/cockroachdb/cockroach/blob/12312fdc0888cd43b553d163b903beabfc391064/pkg/ccl/changefeedccl/changefeed_stmt.go#L968)):*
> ```Go
> // or the TargetSpecifications in new ones.
> func AllTargets(cd jobspb.ChangefeedDetails) (targets []jobspb.ChangefeedTargetSpecification) {
> 	if len(cd.TargetSpecifications) > 0 {
> ```

We need a todo to add a version gate.


<!-- Sent from Reviewable.io -->

HonoreDB reviewed (commented):
Thank you!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75430#-:-MwWz0VM-LcFCTOjTUCH:b-9755fn)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @gh-casper, @HonoreDB, and @miretskiy)

___
*[pkg/ccl/changefeedccl/changefeed_stmt.go, line 968 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/75430#-MwWoFrh47N63DdmBnhQ:-MwWz2yj5O8xYWSXnkdK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/12312fdc0888cd43b553d163b903beabfc391064/pkg/ccl/changefeedccl/changefeed_stmt.go#L968)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

We need a todo to add a version gate.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

HonoreDB commented:
bors r=[miretskiy]
